### ISSUE TITLE
HH3: some sounds, animations, actions (events?) not appearing or appearing incorrectly

### ISSUE THREAD

**Issue by vicusdanielson** [EasyRPG/Player#985]
Greetings! Just as I wrote in the title, some sounds, animations, actions/events? (I meant e.g. the boss's attack animation, boosting himself) do not appear or they appear but incorrectly.
Game name: [Heaven and Hell Episode 3](http://www.anonelbe.net/hh3-rm2k/)
I uploaded 2 videos of my playing (I re-encoded the videos to lower quality for smaller size):
[Smartphone one](https://www.youtube.com/watch?v=bh-O0whvLJo) (where the issues occur) / [Laptop one](https://www.youtube.com/watch?v=XlYl84nxHgk) (which is the correct/how it should be)
Smartphone I played on: [Xiaomi Redmi Note 2](http://www.gsmarena.com/xiaomi_redmi_note_2-6992.php)
Android version: 5.0.2 LRX22G, MIUI version: 7.4 by xiaomi.eu (6.5.19 Developer)

So the things which I noticed to be incorrect on the smartphone one:
- [X] When battles start that beeping sound doesn't get slowed (for some reason it's made like that fast and normally it plays slower - the "pitch" in the rm2k is set to lower)
- [x] On the battle screen the names, etc appear with a "%S" at their end (e.g. Martyr%S attacks the enemy, Kharan 7%S %V damage sustaned)
- [x] The "Animal Summoning III" spell of the common enemies doesn't do its grrr\* sound
- [X] When I "Run away" from battles, it doesn't do its sound and running away is much more successful (maybe I just have better luck)
- [x] "AutoBattle" always does the basic attack, normally it does the strongest attack
- [x] The basic attack does damage one time, normally it does two times **(since 0.6)**
- [x] After I used a spell the next time I attack it doesn't select it automatically when I select Fight/NirvanaSwd (e.g. I used "Sun Flare" spell at 5:10 then next time I attacked at 5:20 it didn't auto-select that spell but always the first one instead) **(since 0.6)**
- [x] When a spell misses on an enemy it doesn't show like "it missed" or "he avoided" (e.g. when the Templar avoids Kharan's "Dead Sword" spell)
- [x] The boss, Abel always just boosts himself (the values of the boosts are incorrect too, e.g. "Attack Pwr -1, ArmorClass 25"), never attacks (I even killed him XD, got the "Debugger" then that battle restarted) **(since 0.6)**
- [x] Shake screen / flash on damage doesn't affect animations. When the enemy uses his physical attack, you can see the last few frames are off as the fake background in the attack animation is offset with the real background. **(since 0.6.1)**
- [x] Blade barrier does not boost the enemies physical element resistances
- [x] The "Steal" skill steals 3 times of the random enemies instead of 1. **(since 0.6.1)**

Please keep up your great work! These issues might appear on different rm2k games.

**Comment by Ghabry** [comment]
Thanks for your detailed bug report, not many are of such quality.

> When battles start that beeping sound doesn't get slowed (for some reason it's made like that fast and normally it plays slower - the "pitch" in the rm2k is set to lower)
Pitch for sound is not implemented yet: Issue #895

> On the battle screen the names, etc appear with a "%S" at their end (e.g. Martyr%S attacks the enemy, Kharan 7%S %V damage sustaned)
The game uses the new RPG Maker 2000 release (English version). We don't fully support this yet. I'm surprised that somebody found a game using it. Issue #588 

> The "Animal Summoning III" spell of the common enemies doesn't do its grrr* sound
Ok

> When I "Run away" from battles, it doesn't do its sound and running away is much more successful (maybe I just have better luck)
Could be luck, we use the same algorithm afaik. The sound is indead missing.

> "AutoBattle" always does the basic attack, normally it does the strongest attack
That's on purpose. Some people are happy that it only does basic attack and now you complain :D. Okay we need some compromise (I hate it when AutoBattle drains my SP)

> The basic attack does damage one time, normally it does two times
Double attack is not implemented yet. Issue #480

> After I used a spell the next time I attack it doesn't select it automatically when I select Fight/NirvanaSwd (e.g. I used "Sun Flare" spell at 5:10 then next time I attacked at 5:20 it didn't auto-select that spell but always the first one instead)
Ok

> When a spell misses on an enemy it doesn't show like "it missed" or "he avoided" (e.g. when the Templar avoids Kharan's "Dead Sword" spell)
Ok

> The boss, Abel always just boosts himself (the values of the boosts are incorrect too, e.g. "Attack Pwr -1, ArmorClass 25"), never attacks (I even killed him XD, got the "Debugger" then that battle restarted)
Could you give us a savegame close to that Boss?

**Comment by vicusdanielson** [comment]
Hiho!

> That's on purpose. Some people are happy that it only does basic attack and now you complain :D.

> Okay we need some compromise (I hate it when AutoBattle drains my SP)

Actually I hate it too :I but then it wouldn't be compromise so..... here's my deal.!?%: I wouldn't really care if the auto-re-select works for skills 83

> Could you give us a savegame close to that Boss?

Unfortunately you cannot save in the "intro", there are no save points (white-filled circle in blue flame). Well, it takes about 5 minutes from "New Game" to get there, like in the videos.

**Comment by vicusdanielson** [comment]
Hello!

The things that are changed in the game in the latest version of the easy rpg now (0.5.0-52) from the last version I reported:

- When battles start that beeping sound doesn't get slowed (for some reason it's made like that fast and normally it plays slower - the "pitch" in the rm2k is set to lower) - _Fixed_
- The boss, Abel always just boosts himself (the values of the boosts are incorrect too, e.g. "Attack Pwr -1, ArmorClass 25"), never attacks (I even killed him XD, got the "Debugger" then that battle restarted) - _This is a bit changed: it's "Attack Pwr 25, ArmorClass 25, Dexterity 25" (dexterity was in the previous version too just forgot to write) and these and every other similar phrases that need also got the "%S %U %V up/down/was taken etc!" at their end, so this got better._
- When I "Run away" from battles, it doesn't do its sound and running away is much more successful (maybe I just have better luck) - _This is not changed but now I know, this is not like I have best luck but running away is not correct. 20/20 I can ran away safely._

I will keep reporting if something changes, even if this is made by the newer Rpg Maker 2000 English release.

**Comment by Ghabry** [comment]
The escape formula is indeed wrong due to some errors in RPG Advocates help file translation. Attached the correct formula.

```cpp
int ally_agi = Main_Data::game_party->GetAverageAgility();
int enemy_agi = Main_Data::game_enemyparty->GetAverageAgility();

float to_hit = 1.5f - ((float)enemy_agi / ally_agi);

// Every failed escape is worth 10% higher escape chance
to_hit += to_hit * Game_Battle::escape_fail_count * 0.1f;

to_hit *= 100;

this->success = Utils::GetRandomNumber(0, 99) < (int)to_hit;
```

Belongs in this if-block:
https://github.com/EasyRPG/Player/blob/master/src/game_battlealgorithm.cpp#L1398

**Comment by carstene1ns** [comment]
Here is a patch to add the escape sound effect. Only tested in the hh3 (2k) game, not 2k3:
```cpp
--- a/src/scene_battle_rpg2k.cpp
+++ b/src/scene_battle_rpg2k.cpp
@@ -641,6 +641,9 @@ void Scene_Battle_Rpg2k::Escape() {
                escape_success = escape_alg.Execute();
                escape_alg.Apply();
 
+               if (escape_success)
+                       Game_System::SePlay(Game_System::GetSystemSE(Game_System::SFX_Escape));
+
                battle_result_messages.clear();
                escape_alg.GetResultMessages(battle_result_messages);
 
diff --git a/src/scene_battle_rpg2k3.cpp b/src/scene_battle_rpg2k3.cpp
index 26a8710a..450925ec 100644
--- a/src/scene_battle_rpg2k3.cpp
+++ b/src/scene_battle_rpg2k3.cpp
@@ -943,6 +943,7 @@ void Scene_Battle_Rpg2k3::Escape() {
                ShowNotification(battle_result_messages[0]);
        }
        else {
+               Game_System::SePlay(Game_System::GetSystemSE(Game_System::SFX_Escape));
                Game_Temp::battle_result = Game_Temp::BattleEscape;
 
                // ToDo: Run away animation
```

**Comment by Ghabry** [comment]
Because Escape inherits from BattleAlgorithm one could also use "GetResultSe()" here, but Escape doesn't overload this yet.

**Comment by carstene1ns** [comment]
This one might have fixes in #1373.

**Comment by fmatthew5876** [comment]
>The boss, Abel always just boosts himself (the values of the boosts are incorrect too, e.g. "Attack Pwr -1, ArmorClass 25"), never attacks (I even killed him XD, got the "Debugger" then that battle restarted)

This battle is very special in that most of the enemy actions entirely scripted with switches and battle events. This was done to give the enemy attack animations which is not normally possible in the DBS.

This enemy has the following configuration.

<img width="233" alt="enemy" src="https://user-images.githubusercontent.com/1198466/43017368-c00b0462-8c4d-11e8-9f95-d8741ba451e0.png">

So according to the bug report, ABEL DRAW UPON is being selected every round instead of only the first round.

As for the boost values being incorrect, that ability is one of the only ones which is not a battle event. Its just a regular buff.

<img width="375" alt="draw" src="https://user-images.githubusercontent.com/1198466/43017510-4e7a0e5a-8c4e-11e8-8945-44aebcd840ed.png">

I am the developer of this game, so if you need more information about any of these please ask. I have not yet tested easyrpg at all myself.

**Comment by fmatthew5876** [comment]
I just tested this with the latest player binary and can report these results:

> The boss, Abel always just boosts himself (the values of the boosts are incorrect too, e.g. "Attack Pwr -1, ArmorClass 25"), never attacks (I even killed him XD, got the "Debugger" then that battle restarted)

I can break this down further.

* In RPG_RT, Abel always casts "Draw Upon Holy Might" only the first turn. In EasyRpg he does one of the other actions on the first turn and always casts DUHM the second turn.
* In RPG_RT, DUHM boosts his 3 stats by 70. in EasyRPG, it only boosts them by 25. (*See below, maybe this is related to multiple stats).
* In RPG_RT, Blade Barrier boosts his defense by 50, and raises his slashing, piercing, and blunt defense. In EasyRPG it only boosts his defense by 50. Maybe it also boosts the resistances but I can't tell as it doesn't show the text indicating so.
* In EasyRPG, none of his "Do Nothing" actions (which turn on switches) work. He just does nothing. I don't know if the switches actually get turned on or not as it appears the F9 debug menu doesn't work in battle.

Btw, this battle can be tested easily in player

`Player64.exe --battle-test 18`

In general the battle system in rm2k mode is much slower. The pacing is off somehow.

> When battles start that beeping sound doesn't get slowed (for some reason it's made like that fast and normally it plays slower - the "pitch" in the rm2k is set to lower)

The battle encounter sound effect is set to tempo 50. It works in EasyRPG if I start a New Game. It still does not work if load a saved game. My saved games were created by RPG_RT, not sure if that makes a difference.

> When I "Run away" from battles, it doesn't do its sound and running away is much more successful (maybe I just have better luck)

This appears to be fixed.

I'm finding lots of other small bugs,, which I'll make a new issue for.

**Comment by carstene1ns** [comment]
We have some battle related pull requests open, might be worth to check after merging.

**Comment by fmatthew5876** [comment]
I tried it with this build:
* https://ci.easyrpg.org/view/Pull%20Requests/job/player-win32-pr/1698/

The only thing it fixes is that DUHM now correctly boosts Abel's stats by 70. Also the general slowness of rm2k battles is fixed which is great!

**Comment by carstene1ns** [comment]
(ticked sound and message checkboxes in original issue description)

**Comment by fmatthew5876** [comment]
I did some more testing here. All of the original issues are fixed (except autobattle, but thats a separate topic).

However, there are still more problems with this battle

- [ ] Shake screen on damage doesn't shake animations. When the enemy uses his physical attack, you can see the last few frames are off as the fake background in the attack animation is offset with the real background.
- [ ] Blade barrier does not boost the enemies physical element resistances

**Comment by vicusdanielson** [comment]
Event doesn't progress at the boss battle in the SkyFlow building after Sirran says "Destroy these creatures for me.. They are no longer needed!"

I uploaded the save file (works only on android). Just walk north to the next area then walk north till you trigger this event.
[Save01.zip](https://github.com/EasyRPG/Player/files/3002031/Save01.zip)

**Comment by fmatthew5876** [comment]
The steal 3 times issue is surely #1652

I will look at the skyflow thing. That sounds event/move route related in which case it may already be fixed by one of the outstanding PRs

**Comment by fmatthew5876** [comment]
I've confirmed the game does hang during the second boss battle. This is caused by #1612 which is already fixed by #1687.

Thanks for the bug report!

**Comment by fmatthew5876** [comment]
Everything here should now be fixed in the continuous build except the flash screen issues in the intro battle.

**Comment by fmatthew5876** [comment]
Unfortunately, this is still broken. The screen flash is off by 1 frame in the battle PR

![frame00000074](https://user-images.githubusercontent.com/1198466/58677690-03accf00-832b-11e9-9858-48b5b62fdb99.png)
![frame00000075](https://user-images.githubusercontent.com/1198466/58677691-04456580-832b-11e9-8077-f80160faf7f8.png)
![frame00000076](https://user-images.githubusercontent.com/1198466/58677692-04456580-832b-11e9-90fc-0f7904f7012d.png)

**Comment by fmatthew5876** [comment]
This is fixed for real by #1788
